utils/zerovec/src/ule/custom/encode.rs

```diff
@@ -100,11 +100,23 @@ unsafe impl EncodeAsVarULE<str> for String {
     }
 }
 
-unsafe impl<T: ULE, A: AsULE<ULE = T> + 'static> EncodeAsVarULE<ZeroVecULE<A>> for Vec<T> {
-    fn encode_var_ule_as_slices<R>(&self, cb: impl FnOnce(&[&[u8]]) -> R) -> R {
-        cb(&[<ZeroVecULE<A> as VarULE>::as_byte_slice(
-            ZeroVecULE::<A>::from_slice(self),
-        )])
+unsafe impl<T> EncodeAsVarULE<ZeroVecULE<T>> for Vec<T>
+where
+    T: EqULE + 'static
+{
+    fn encode_var_ule_as_slices<R>(&self, _: impl FnOnce(&[&[u8]]) -> R) -> R {
+        // unnecessary if the other two are implemented
+        unreachable!()
+    }
+
+    fn encode_var_ule_len(&self) -> usize {
+        self.len() * core::mem::size_of::<T>()
+    }
+
+    fn encode_var_ule_write(&self, dst: &mut [u8]) {
+        let zv = ZeroVec::from_slice(&self);
```

Comment:
issue: we can make this work in an unconditionally zero-copy way as well; by looping over the vector, calling `T::as_unaligned()`, and appending. The main difference is that that loop is a little more complex (here you have a straightforward memcopy)

Overall EncodeAsVarULE is most of the time called before serialization, so I'm not sure if we should be trading off API flexibility for a minor performance win.

Perhaps we should remove the EqULE bound and use a for loop here instead?

Reply:
Yeah, I was thinking about these things and went back-and-forth a bit.

There are several ways we could do this:

1. Bound on `T: ULE, T: AsULE<ULE = T>` like before
2. Bound on `T: EqULE` and use `ZeroVec::from_slice` (cheap on LE, allocation required on BE)
3. Bound on `T: EqULE` and use `ZeroVec::try_from_slice`, falling back to a loop over each element
4. Bound on `T: AsULE` and always loop over each element
5. Bound on `T: AsULE` as above, with an optimization that checks `TypeId::of::<T>() == TypeIf::of::<T::ULE>()` and then transmutes and runs a normal copy

This PR moves away from (1) because it is too limiting.  We should be able to convert `Vec<u32>` into `ZeroVecULE<u32>`.

(2) is easy and covers our main use case on LE in an efficient way.

I prefer not to do (3) because it introduces a code path that is hard to test and rarely gets used (a BE-only code path).

(4) is doable, but more complicated to write, and is likely slower than (2) or (3) on LE because `memcpy` is faster than an iterator

(5) is also doable, but not as fast as (2) or (3) on LE because we can only check for `T == T::ULE` at runtime, not `T: EqULE` (I think??)

Reply:
I wish we could use specialization and obtain the following table of behavior:

| Type | Example | LE | BE |
|---|---|---|---|
| T == T::ULE | `u8` | memcpy | memcpy |
| T: EqULE | `u32` | memcpy | loop |
| T: AsULE | `PatternItem` | loop | loop |

Reply:
So the thing that makes me prefer `T: AsULE` here is that EncodeAsVarULE is to me not one of the things that is in zerovec "fast path" bucket. zerovec is optimized for zero-copy deserialization and _reads_, writes need not be efficient since they are either:

 - Used for constructing things for serialization, which is itself not in the "fast path" bucket
 - Used very occasionally for "patching" data

So to me EncodeAsVarULE is something we should _try_ to optimize but if there's a tradeoff between performance (here, using loops vs memcpy for LE) and API flexibility I'd rather pick API flexibility.

Ideally we'd be able to implement on both `T: AsULE` and `T: ULE` but unfortunately this leads to conflicting impls.

Reply:
My compromise is that `VZV<[T]>` uses the fast path and is only defined for `T == T::ULE`, and `VZV<ZeroSlice<T>>` uses the "slower" path.